Scoob

stefi07 opened the pull request:
As discussed in Slack, adding an option to be configured in a config file to display either the MagAO-X or the Scoob GUI.

The first two commits are more generic, adding some missing python dependencies and standardizing component calls.

The last commit makes the changes necessary to read the config file and switch between different GUI layouts:
- read 'layout' string from '/opt/MagAOX/config/gui.conf with xconf (full file path can be passed in as variable); options for 'layout' are currently 'MAGAOX' or 'SCOOB'; default is 'MAGAOX'.
- query 'layout' value in Vue via axios (common.js).
- move existing files in frontend/pages to frontend/pages/MagAOX and created fontend/pages/Scoob.
- move tabs templates into frontend/pages/MagAOX/TabBar.vue & create frontend/pages/Scoob/TabBar.vue.
- create fontend/components/basic/TabBar.vue; depending on the 'layout' value, fontend/components/basic/TabBar.vue loads either frontend/pages/MagAOX/TabBar.vue or frontend/pages/Scoob/TabBar.vue.

stefi07 committed:
Added missing dependencies to setup.py.

stefi07 committed:
Standardizing component imports to be relative to Vue parent folder and not relative to importing file.

stefi07 committed:
Reading 'layout' string from '/opt/MagAOX/config/gui.conf with xconf (full file path can be passed in as variable).
Options for 'layout' are currently 'MAGAOX' or 'SCOOB'. Default is 'MAGAOX'.
Vue queries 'layout' value via axios (common.js).
Moved existing files in frontend/pages to frontend/pages/MagAOX and created fontend/pages/Scoob.
Moved tabs templates into frontend/pages/MagAOX/TabBar.vue. Created frontend/pages/Scoob/TabBar.vue.
Created fontend/components/basic/TabBar.vue. Based on 'layout' value, fontend/components/basic/TabBar.vue loads
either frontend/pages/MagAOX/TabBar.vue or frontend/pages/Scoob/TabBar.vue.

stefi07 committed:
Collecting all hardcoded constants in constants.json

Constants.json also contains defaults for the options that can be
configured in the config file.
The backend & frontend read constants.json and extract the constants
they need in constants.py and constants.js respectively.

Added replicated_cameras as an extra config option.
As in the previous commit, core.py parses the config file. Any
constants defined in the config file replace the constants defined in
constants.json.

Also fix on constants import in common.js

stefi07 commented:
I've added another commit that collects all hardcoded constants in constants.json
Constants.json also contains defaults for the options that can be configured in the config file.
The backend & frontend read constants.json and extract the constants they need in constants.py and constants.js, respectively.

Added REPLICATED_CAMERAS as an extra config option. As in the previous commit, core.py parses the config file. Any constants defined in the config file replace the constants defined in constants.json.

jaredmales commented:
I recommend making the config file `gui_sup.conf` to conform to the convention already established.  See, e.g., `gui_camsci1.conf`

joseph-long commented on frontend/common.js:
```diff
@@ -134,6 +141,32 @@ export default {
           Vue.set(this.indiWorld[deviceName], propertyName, state);
         },
+        fetchRoutes() {
+          this.$axios.get(constants.CONFIG_URL)
```
I think we can skip adding axios and just use the fetch api directly here

joseph-long replied:
here's an example: https://github.com/magao-x/sup/blob/main/frontend/components/plots/ObservabilityPlots.vue#L97

stefi07 replied:
Sure, that makes sense, I've refactored to use fetch instead. I've also moved the call in index.js so that the relevant routes are loaded from the beginning and changed to switched to storing config data in store which removes the need to inject it in elements.

joseph-long commented on frontend/components/basic/TabBar.vue:
```diff
@@ -0,0 +1,28 @@
+    TabBarScoob,
+  },
+  computed: {
+    showScoob: function () {
```
Since we're refactoring, we might as well future-proof for the Nth instrument to use this UI... instead of a boolean yes-scoob/no-scoob we should test for equality in the `v-if` in the template and have a fallback for if an unknown layout is requested that says something useful (to catch misconfigurations). Then the computed property won't need to be renamed when support for a third instrument is added

stefi07 replied:
Very good point, I've made the change.

joseph-long commented on sup/constants.py:
```diff
@@ -1,11 +1,20 @@
+
+import json
+
+with open('../constants.json', 'r') as file:
```
As written, this isn't working... `../` is relative to the sup process current working directory, which we don't expect to be `/opt/MagAOX/source/sup/sup`. Also, we always install `sup` in editable mode, but I don't think we should depend on that being true here.

stefi07 replied:
Yes, that's a good point. I can see three solutions:

1. Assert separation of concerns and go back to decoupling front and back end constants. Put any constants used by both in the config file. Default values for config file options are set in the backend constants file and the frontend assumes it will always receive a value.
2. Move the consolidated json constants file in the backend, read it there and pass all variables required by the frontend through the fetch call.
3. Move all constants in the config file. The backend reads it and passes all variables required by the frontend through the fetch call. This would need to enforce a minimum level of config present in the config file, since there wouldn't be any default values.

Do you have a preferred solution or a convention that you're already using? 
I think there's no ideal solution and it depends on what we're trying to optimize. I'm leaning towards the first approach, going back to two constants files.

joseph-long replied:
I agree there's no perfect solution, but I favor option 1. If that requires too many onerous gymnastics to get the frontend configured, we can also just live with redundancy for now and use elbow grease to keep two constants files in sync. We don't mix and match frontend and backend versions, so they should evolve in sync.

stefi07 replied:
Cool, I've made these changes.

I considered dynamically building the tab bar import paths from the layout variable, which is why I converted the config values and folder names to lowercase ('magaox' & 'scoob'). However that would require including Webpack as a dependency and we wouldn't fully eliminate manual input anyway, since the routes would still need to be mapped manually to the corresponding layout (what is now in map.js). In the end it doesn't seem worth the effort at this stage, but it's still an option in the future. But I kept the lowercase convention to have it all self-consistent.

joseph-long commented on setup.py:
```diff
@@ -75,12 +75,18 @@ def run(self):
         'matplotlib>=3.5.1',
         'uvloop>=0.16',
-        'purepyindi2>=0.0.0',
+        'purepyindi2 @ git+http://git@github.com/xwcl/purepyindi2.git',
```
I'm not sure that listing the git URLs here is good. It's better for redistribution, but all of these evolve with MagAO-X software and at least one of them is nontrivial to install (depends on the version of MILK used in the control loop). While this is under heavy development I think all we should do is list them to make sure they exist on the package search path, but not trust pip to install them correctly automatically

stefi07 replied:
:) it was an easy way to set up my dev env, but I see your point. I reverted to the non-git requirements list.

joseph-long commented on sup/utils.py:
```diff
@@ -1,9 +1,23 @@
+    replicated_cameras : list[str] = xconf.field(help="List of cameras", default_factory=lambda: REPLICATED_CAMERAS)
+
+    def main(self):
+        breakpoint()
```
leftover?

stefi07 replied:
Indeed. Removed.

joseph-long commented on sup/utils.py:
```diff
@@ -1,9 +1,23 @@
+
+@xconf.config
+class GUIConfig(xconf.Command):
+    """Demo command"""
```
Leftover?

stefi07 replied:
Indeed. Updated.

joseph-long commented on sup/core.py:
```diff
@@ -87,6 +95,18 @@ def log_updates(update, did_anything_change):
 
 sup_tasks = []
 
+def parse_config_file():
```
I may contribute some edits of my own here since I have ideas about how `xconf` "should" work, but as long as it's working it's fine for now.

I'm not sure it's necessary to re-read the config every page-load, though actual complete reloads happen relatively infrequently so it's not like it's a load on the system. I'm just thinking the way to get everything to take effect after editing a config file should be to restart the service itself, for consistency.

stefi07 replied:
I removed the config re-read on page-load and left it only on service initialization. Happy (and curious) for any edits needed to standardize the use of `xconf`

joseph-long reviewed (changes requested):
It's great to have some help cleaning this codebase up! It's kind of like finding out someone tidied up your apartment while you were out. This project mostly gets added to around 4am at a telescope, so normalizing e.g. import paths was not a priority.

I identified a couple of issues to do with how we actually deploy sup on the instrument (paths, packaging, etc) where I left comments, but if I hard-code the config.json path in I can start the software from your branch in-situ on MagAO-X, so this is a great start.

stefi07 committed:
Addressing PR #17 comments.

* Updated default config file name to gui_sup.conf.
* Refactored frontend config request to use fetch API. Also moved this
call into index.js and switch storing config data in store which removes
the need to inject it in elements.
* Refactored layout check in TabBar.vue to be general.
* Removed git links in setup.py.
* Removed config file check on page load. Now the config file is only
parsed on service initialization.
* Cleaned out leftover bits.

Outstanding: where to store non-config constants. Unchanged for now.

stefi07 committed:
Added error message in frontend/components/basic/TabBar.vue if the
'layout' value is invalid.

stefi07 committed:
Reverted to separate constant files for backend & frontend.

Corresponding routes and tabBar elements are mapped for each layout
type in map.js to keep all manual configuration in one place.
An error is shown if no corresponding TabBar element exists for the
layout.

joseph-long commented:
Thanks for your work, @stefi07 ! As you will see when you pull from `main` next, I've made some additional changes. There was a pretty drastic refactoring of the Python backend, because it was (ab-)using Flask-style module-level variables as globals. The entry point has changed, so if you did `pip install -e .` to install before, you'll have to do it again to get the `sup` command working again.
